src/tls_aead/stream_sink.rs

```diff
@@ -0,0 +1,156 @@
+// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::{
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::Poll,
+};
+
+use futures::{ready, Sink, Stream};
+use thiserror::Error;
+use tokio_util::bytes::{Bytes, BytesMut};
+
+use crate::tls_aead::{codec::TlsPacketIn, SecretUpdate};
+
+use super::{TlsAeadCodec, TlsAeadCodecError};
+
+#[derive(Debug, Error)]
+pub enum TlsAeadSinkError {
+    #[error(transparent)]
+    EncryptionError(#[from] TlsAeadCodecError),
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+    #[error("Poisoned lock error")]
+    PoisonedLockError,
+}
+
+pub(crate) struct TlsAeadSink<S> {
+    inner: S,
+    cipher: Arc<Mutex<TlsAeadCodec>>,
+}
+
+impl<S> TlsAeadSink<S> {
+    pub(crate) fn new(inner: S, cipher: Arc<Mutex<TlsAeadCodec>>) -> Self {
+        Self { inner, cipher }
+    }
+
+    pub(crate) fn update_traffic_secrets(
+        &mut self,
+        traffic_secrets: SecretUpdate,
+        is_server: bool,
+    ) -> Result<(), TlsAeadSinkError> {
+        self.cipher
+            .lock()
+            .map_err(|_| TlsAeadSinkError::PoisonedLockError)?
+            .update_traffic_secrets(traffic_secrets, is_server)?;
+        Ok(())
+    }
+}
+
+impl<S> Sink<Bytes> for TlsAeadSink<S>
+where
+    S: Sink<Bytes, Error = std::io::Error> + Unpin,
+{
+    type Error = TlsAeadSinkError;
+
+    fn poll_ready(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        Pin::new(&mut self.get_mut().inner)
+            .poll_ready(cx)
+            .map_err(From::from)
+    }
+
+    fn start_send(self: std::pin::Pin<&mut Self>, item: Bytes) -> Result<(), Self::Error> {
+        let sink = self.get_mut();
+        let mut cipher = sink
+            .cipher
+            .lock()
+            .map_err(|_| TlsAeadSinkError::PoisonedLockError)?;
+
+        let ciphertexts = cipher.encrypt(&item)?;
+        drop(cipher); // Release the lock before sending
+        for ciphertext in ciphertexts {
+            Pin::new(&mut sink.inner).start_send(ciphertext.into())?;
```

Comment:
If I am not mistaken, the API of `Sink` requires first to call `poll_ready` on the inner sink, before calling start_send: https://docs.rs/futures/latest/futures/prelude/trait.Sink.html#tymethod.start_send

In particular this means, the `ciphertexts` must be stored in a buffer.